enhancements/20220530-vxlan-enhancement.md

```diff
@@ -0,0 +1,521 @@
+
+`FDB`: Forwarding Database
+
+## Motivation
+
+With the enhancement of Harvester VLAN network, user can deploy up to 4K isolated `VM group` in one `Host Harvester Cluster`.
+
+![](./20220530-vxlan-enhancement/harvester-vlan-network-1.png)
+
+ Meanwhile, it means that the physical network needs to prepare up to 4K VLAN for `Host Harvester Cluster`. This is a challenge for massive deployment of `Guest Kubernetes Cluster` .
+
+ When user plans to deploy more than 4K isolated guest clusters, it is not possible now.
+
+ Harvester will try to eliminate those restrictions in this enhancement.
+
+![](./20220530-vxlan-enhancement/harvester-vxlan-network-1.png)
+
+### Goals
+
+Add VXLAN network with adding the following advanced features:
+
+  (3) Support up to 16K isolated `VM Group` in one `Host Harvester Cluster`, VM in different group can not communicate with each other direclty.
+
+  (3) Support VM live migration and persistent IP address, VM can freely migrate/move in the cluster without change of IP address. 
+
+  (3) Eliminate the requirements of VLAN in physical network, as few as one VLAN id . 
+
+  (3) Support automatically establishing of VXLAN tunnel, all NODEs in a dedicated VLAN physical network, should be able to build full mesh VXLAN tunnel connection.
+
+  (1) Support pure L3 IP network in physical network ? (is not supported yet, in future, it is possible)
+
+  (3) Support VM communicate with VM in same host NODE, in different host NODE
+
+  (3) Support VM communicate with outside (e.g. internet)
+
+  TBD: Support network policies for micro-segmentation of workloads [FEATURE] Support network policies for micro-segmentation of workloads #2260 .
+
+  (2) Support multiple VXLANs with separated physical NICs.
+
+  (3) Integrate with kube-vip to provide load balancer service for VMs based on vxlan network.
+
+
+NOTE: (3) is priority, bigger one is more important.
+
+[quck view of VXLAN user operation](#an-example-of-vxlan-user-operation)
+
+### Non-goals [optional]
+
+.
+
+## Proposal
+.
+
+### User Stories
+
+.
+
+#### Story 1
+
+VXLAN tunnel establishment
+
+![](./20220530-vxlan-enhancement/vxlan-vtep-1.png)
+
+For VXLAN network, it is trivial to add/remove peers in each node manually, even with kind of centralized registry, the HA will still be a challenge.
+
+But at the view of Harvester, those drawbacks are not here, each NODE has all the required information to establish VXLAN tunnels.
+
+#### Story 2
+
+Adding Virtual Network 
+
+![](./20220530-vxlan-enhancement/vxlan-vni-1.png)
+
+User defines (cluster level) VN and the related CIDR, then the VNI, VXLAN tunnel are created automatically.
+
+#### Story 3
+
+Adding VM Group into a Virtual Network
+
+![](./20220530-vxlan-enhancement/vxlan-vmg-1.png)
+
+Again, Harvester Host Cluster has knowledge of all the existing VMs, it can convert them into related FDB, the VTEP can skip the dynamic MAC learning.
```

Comment:
It's a kind of SDN controller.
Do we want to have a SDN controller for this?
It would be similar to the OpenStack neutron implementation